llvm/lib/Transforms/Yk/ShadowStack.cpp

```diff
@@ -271,6 +276,62 @@ class YkShadowStack : public ModulePass {
     }
   }
 
+  // Handle main() separatley, since it works slightly differently to other
+  // functions: it allocates the shadow stack
+  //
+  // For main clone: `__yk_unopt_main` we need to update the allocas
+  // while reusing the same shadow stack pointer as the original main.
+  //
+  // Note that since we assuming main() doesn't call main(), we can consider
+  // the shadow stack disused at the point main() returns. For this reason,
+  // there's no need to emit a shadow epilogue for main().
+  //
+  // YKFIXME: Investigate languages that don't have/use main as the first
+  // entry point.
+  void updateMainFunctions(DataLayout &DL, Module &M, GlobalVariable *SSGlobal,
+                           LLVMContext &Context) {
+    Function *Main = M.getFunction(MAIN);
+    if (Main == nullptr || Main->isDeclaration()) {
+      Context.emitError("Unable to add shadow stack: could not find definition "
+                        "of \"main\" function!");
+      return;
+    }
+
+    // Update the original main function.
+    std::map<AllocaInst *, size_t> MainAllocas;
+    std::vector<ReturnInst *> MainRets;
+    size_t SFrameSize = analyseFunction(*Main, DL, MainAllocas, MainRets);
+    CallInst *Malloc = insertMainPrologue(Main, SSGlobal, SFrameSize);
+    rewriteAllocas(DL, MainAllocas, Malloc);
+
+    // If we have two control points, we need to update the cloned main
+    // function as well.
+    if (controlPointCount == 2) {
+      Function *MainClone = M.getFunction(YK_UNOPT_MAIN);
+      if (MainClone == nullptr || MainClone->isDeclaration()) {
+        Context.emitError(
+            "Unable to add shadow stack: could not find definition "
+            "of \"__yk_unopt_main\" function!");
+        return;
+      }
+
+      std::map<AllocaInst *, size_t> MainCloneAllocas;
+      std::vector<ReturnInst *> MainCloneRets;
+      size_t SFrameSizeClone =
+          analyseFunction(*MainClone, DL, MainCloneAllocas, MainCloneRets);
+
+      // Insert a load of SSGlobal at the beginning of MainClone.
+      BasicBlock &EntryBB = MainClone->getEntryBlock();
+      Instruction *FirstNonPHI = EntryBB.getFirstNonPHI();
+      IRBuilder<> BuilderClone(FirstNonPHI);
+
+      // Load the current shadow stack pointer from the global variable.
+      LoadInst *LoadedSSPtr = BuilderClone.CreateLoad(Int8PtrTy, SSGlobal);
+      // Replace all allocas in MainClone with the loaded shadow stack pointer.
+      rewriteAllocas(DL, MainCloneAllocas, LoadedSSPtr);
```

Comment:
We need at the very least a comment explaining why we believe that allocas in unopt will match the allocas in opt. But ideally we have asserts here that make sure. We might want to optime the `opt` version slightly in the future which could change the variable order and then things will go wrong.